_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,468 @@
+[gpt]:https://openai.com/blog/language-unsupervised/#drawbacks
+[avoiding repeated work]:https://ml.dask.org/hyper-parameter-search.html#avoid-repeated-work
+[BOAH]:https://github.com/automl/BOAH
+[Auto-PyTorch]:https://www.automl.org/wp-content/uploads/2018/09/chapter7-autonet.pdf
+[HpBandSter]:https://github.com/automl/HpBandSter
+[blippar]:https://arxiv.org/pdf/1801.01596.pdf
+[BOHB paper]:http://proceedings.mlr.press/v80/falkner18a/falkner18a.pdf
+[n-grams]:https://en.wikipedia.org/wiki/N-gram
+[^smac]:SMAC is described in "[Sequential Model-Based Optimization forGeneral Algorithm Configuration][SMAC-paper]," and is available [in AutoML][SMAC-automl].
+[^spearmint]:Spearmint is described in "[Practical Bayesian Optimization of MachineLearning Algorithms][pbo]," and is available in [HIPS/spearmint].
+[^tpe]:TPE is described in Section 4 of "[Algorithms for Hyperparameter Optimization][aho]," and is available [through Hyperopt][hyperopt].
+[SMAC-paper]:https://www.cs.ubc.ca/~hutter/papers/10-TR-SMAC.pdf
+[SMAC-automl]:https://www.automl.org/automated-algorithm-design/algorithm-configuration/smac/
+[spearmint]:https://github.com/HIPS/Spearmint
+[aho]:http://papers.nips.cc/paper/4443-algorithms-for-hyper-parameter-optimization.pdf
+[pbo]:https://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf
+[HIPS/spearmint]:https://github.com/HIPS/Spearmint
+[hyperopt]:http://hyperopt.github.io/hyperopt/
+
+[^stopping]:Hyperband's theory answers "how many models should be stopped?" and "when should they be stopped?"
+[nni]:https://nni.readthedocs.io/en/latest/Tuner/HyperbandAdvisor.html
+[rt-hb]:https://docs.ray.io/en/master/tune/api_docs/schedulers.html#asha-tune-schedulers-ashascheduler
+[rts-hb]:https://docs.ray.io/en/master/tune/api_docs/sklearn.html
+
+### Framework support
+
+*Dask-ML model selection supports libraries like Scikit-Learn, PyTorch, Keras, LightGBM and XGBoost.*
+
+Ray's tune-sklearn supports these frameworks:
+
+> tune-sklearn is used primarily for tuning
+> Scikit-Learn models, but it also supports and provides examples for many
+> other frameworks with Scikit-Learn wrappers such as Skorch (Pytorch),
+> KerasClassifiers (Keras), and XGBoostClassifiers (XGBoost).
+
+However, both Dask-ML and Ray have some qualifications. Certain libraries don't
+offer an implementation of `partial_fit`,[^ray-pf] so not all of the modern
+hyperparameter optimization techniques can be offered. That means Bayesian
+sampling will have to be relied upon, which gets less and less relevant as the
+number of workers grows. Here's a table comparing different libraries and their
+support in Dask-ML's model selection and Ray's tune-sklearn:
+
+| Model Library | Dask-ML support | Ray support | Dask-ML: early stopping? | Ray: early stopping? |
+|:-----:|:-----:|:-----:|:-----:|:-----:|
+| [Scikit-Learn] | ✔ | ✔ | ✔\* |✔\* |
+| [PyTorch] (via [Skorch]) | ✔ | ✔ | ✔ |✔ |
+| [Keras] (via [SciKeras]) | ✔ |✔ | ✔\*\*| ✔\*\* |
```

Comment:
Is Ray also working one this?

Reply:
No. But the work around https://github.com/adriangb/scikeras/issues/24 means that Ray can use it too. I've changed the footnote:

> ** Thanks to work by the Dask developers around [scikeras#24][1]

[1]:https://github.com/adriangb/scikeras/issues/24